modules/job-info: New module

chu11 opened the pull request:
After several WIPs (#2066, #2065), I think this PR is a candidate for merger.

At this moment, the module only supports getting/watching eventlog entries by jobid.  Included are two new job subcommands, `flux job eventlog` and `flux job wait-event`.  They do about what you think they would do.

Adjusted tests that previously used `flux kvs eventlog get` to use new subcommands, removing the need to do `flux job id` commands to get paths.  `flux job wait-event` conveniently removes a big function in `t2300-sched-simple.t` that emulated the same exact thing (Edit: after rebase, also in t2202-job-manager.t`).

I had begun to add a proxy to ensure guest access to the eventlog was possible, but when I tried to add tests, it wasn't clear to me how one submits a job as a guest user, or if I have to fake it in some way right now.  In addition, a number of the functions in `job-manager/util.[ch]` should probably be exported for convenience first.  So I decided perhaps that could be in a follow up PR.

garlick commented on src/modules/job-info/job-info.c:
```diff
@@ -0,0 +1,411 @@
+#include <jansson.h>
+#include <flux/core.h>
+
+#include "src/common/libjob/job.h"
```
Included by <flux/core.h>, so redundant.

garlick commented on src/modules/job-info/job-info.c:
```diff
@@ -0,0 +1,411 @@
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        if (ctx->lookups)
+            zlist_destroy (&ctx->lookups);
```
Any in progress lookups should probably be canceled here (and removed from list and destroyed).

chu11 replied:
I've set a free function on the zlist entries so they are freed via that mechanism.  But you're right, should cancel any watches.

garlick commented on src/common/libjob/job_info.h:
```diff
@@ -0,0 +1,36 @@
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#ifndef _FLUX_CORE_JOB_EVENTLOG_H
```
I forgot where we left the discussion, but I think @grondo earlier proposed just adding this to job.h rather than creating job_info.h, job_types.h, etc., just to keep the number of installed headers under control.

chu11 replied:
That was discussion on `job_util.h`.  I felt `job_info.h` warranted a new file given future additions?  But perhaps we are thinking too far ahead given your comment below on what the API should look like.  I'll just squash it all into `job.h` for the time being.

garlick commented on src/common/libjob/job_info.h:
```diff
@@ -0,0 +1,36 @@
+#endif
+
+flux_future_t *flux_job_eventlog_lookup (flux_t *h, int flags, flux_jobid_t id);
+int flux_job_eventlog_lookup_get (flux_future_t *f, const char **event);
```
I wonder if we could actually drop the flags and rename to `flux_job_event_watch()`?  I was plannig to add a `flux_job_state_watch()` that would return one state per response, so this could be kind of paired with that?

We'll need a function for fetching static job info in bulk (R, jobspec, J, etc).  Maybe that one could be used to fetch the eventlog in its entirety?

I'm not at all sure what the right interfaces are here, so one possibility is to go with what you have and tweak later.

grondo replied:
Agreed, I think exported interfaces in `libjob` related to job "events" could hide the fact from (most) users that the events are in an event log. I like the parity with `flux_job_state_watch`.

I'm also fine starting with what you have though. Probably more important are the cmdline utilities and Python interface...

chu11 replied:
So are you thinking `flux_job_event_watch()`, `flux_job_event_watch_get()`, and `flux_job_event_watch_cancel()`?  Sounds like a good idea if there will be other similar ones.

> We'll need a function for fetching static job info in bulk (R, jobspec, J, etc). Maybe that one could be used to fetch the eventlog in its entirety?

Yeah, once that API / interface becomes more clear, I think that could easily be used for the full eventlog retrieval.

grondo commented:
>  I'm not sure if there's something racy that's possible?

Maybe  a `flux job wait-event --timeout=5 priority` is required before dumping the eventlog?

garlick commented:
Yup that sounds right.  To be clear, that race was not introduced in this PR.

chu11 commented:
Just re-pushed with changes noted above.  Added new `flux_job_event_watch(), flux_job_event_watch_get(), flux_job_event_watch_cancel()` functions.  

`flux_job_eventlog_lookup/lookup_get()` is only for retrieving the full eventlog now.  In the future, there could be a more generic "get R, get J, etc." bulk API functions but decided to leave this here for now.

garlick commented:
one builder failed the same way as reported above

```
expecting success: 
	jobid=$(tail -1 <list10_ids.out) &&
	flux job eventlog $jobid \
		| cut -d" " -f2- | grep ^priority >pri.out &&
	grep -q priority=31 pri.out
not ok 23 - job-manager: priority was updated in KVS
```

I think @grondo's suggested fix ought to solve this e.g.
```patch
diff --git a/t/t2202-job-manager.t b/t/t2202-job-manager.t
index b33b2fac4..76b24efd0 100755
--- a/t/t2202-job-manager.t
+++ b/t/t2202-job-manager.t
@@ -153,6 +153,7 @@ test_expect_success 'job-manager: flux job priority sets last job priority=31' '
 
 test_expect_success 'job-manager: priority was updated in KVS' '
        jobid=$(tail -1 <list10_ids.out) &&
+       flux job wait-event --timeout=5.0 ${jobid} priority >pri.out &&
        flux job eventlog $jobid \
                | cut -d" " -f2- | grep ^priority >pri.out &&
        grep -q priority=31 pri.out
```

chu11 commented:
was going to fix that in a different PR, but I'll add it in this one.

chu11 commented:
missed a couple paths on diff coverage, let me add more tests

chu11 commented:
@garlick I just realized, I didn't add one feature you requested.  That after a single event in 'inactive' the watcher should exit.  I'll add that too.

grondo commented on t/t2204-job-info.t:
```diff
@@ -0,0 +1,109 @@
+test_expect_success NO_CHAIN_LINT 'flux job wait-event hangs on no event' '
+        jobid=$(flux jobspec --format json srun -N1 hostname | flux job submit)
+        ! run_timeout 3 flux job wait-event $jobid foobar
+'
```
The 3s timeout here seems excessive, maybe just left over debug value?

In fact shouldn't .5 or even .2 seconds be enough? (run_timeout supports < 1s timeouts)  then the test overall will run 3x faster.

chu11 replied:
I simply wanted a number that felt "timeout like", but you're right, it need not be larger.

grondo replied:
Yeah, figured it was something like that. The drawback is that we *always* pause for the timeout there.. so we want it to be as long as necessary and no longer ;-)

grondo commented on t/t2204-job-info.t:
```diff
@@ -0,0 +1,109 @@
+#
+
+test_expect_success 'flux job eventlog works (active)' '
+        jobid=$(flux jobspec --format json srun -N1 hostname | flux job submit) &&
```
Suggestion: run `flux jobspec` once and save the output to a file for use in subsequent tests (see e.g. `t2202-job-manager.t`. This will slightly speed up the test.

grondo commented on src/modules/job-info/job-info.c:
```diff
@@ -0,0 +1,420 @@
+            goto done;
+        }
+        else
+            flux_log_error (ctx->h, "%s: flux_kvs_lookup_get", __FUNCTION__);
```
Maybe this error should be supressed (at least for ENOENT), since it is likely a result of looking up an invalid jobid? (No reason to spam the logs, since the error is returned to the user anyway)

grondo commented on src/cmd/flux-job.c:
```diff
@@ -603,6 +626,175 @@ int cmd_id (optparse_t *p, int argc, char **argv)
+    }
+
+    if (flux_job_event_watch_get (f, &s) < 0)
+        log_err_exit ("flux_job_event_watch_get");
```
Suggestion, on at least ENOENT, return a more readable error to the user, e.g. "job %lu not found" instead of `flux_job_event_watch_get: No such file or directory`

grondo commented on src/cmd/flux-job.c:
```diff
@@ -603,6 +626,175 @@ int cmd_id (optparse_t *p, int argc, char **argv)
+    }
+    id = strtoull (argv[optindex++], &endptr, 10);
+    if (*endptr != '\0')
+        log_err_exit ("id requires an unsigned integer argument");
```
Switch to `log_msg_exit()` here to suppress a possibly confusing `strerror (errno)` appended to the end of the error, i.e. I'm getting:
```
flux-job: id requires an unsigned integer argument: No such file or directory
```

grondo commented on src/cmd/flux-job.c:
```diff
@@ -603,6 +626,175 @@ int cmd_id (optparse_t *p, int argc, char **argv)
+    if (flux_rpc_get (f, NULL) < 0) {
+        if (errno == ETIMEDOUT) {
+            fprintf (stderr, "wait-event timeout on event '%s'\n",
+                     ctx->wait_event);
```
should this use `log_msg()` to get command prefix as in other error messages?

grondo commented on src/cmd/flux-job.c:
```diff
@@ -603,6 +626,175 @@ int cmd_id (optparse_t *p, int argc, char **argv)
+    }
+    id = strtoull (argv[optindex++], &endptr, 10);
+    if (*endptr != '\0')
+        log_err_exit ("id requires an unsigned integer argument");
```
`s/log_err_exit/log_msg_exit`

chu11 commented:
- re-pushed, fixed up things commented on by review by @grondo 
- added more tests for hopefully more coverage
- if a lookup reaches the 'inactive' directory of eventlog, assume no more events will occur and end watching.

May be worth noting that due to the latter point above, some of the tests can be racy with ```flux kvs move <activedir> <inactivedir>```, so I worked around it by doing:

```
        activekvsdir=$(flux job id --to=kvs-active $jobid) &&
        inactivekvsdir=$(echo $activekvsdir | sed 's/active/inactive/') &&
        flux kvs copy ${activekvsdir}.eventlog ${inactivekvsdir}.eventlog &&
        flux kvs eventlog append ${inactivekvsdir}.eventlog foobar &&
        flux kvs unlink ${activekvsdir}.eventlog &&
```

chu11 commented:
was debating if this code is correct when the module is being unloaded

```
            while ((l = zlist_pop (ctx->lookups))) {
                if ((l->flags & FLUX_JOB_INFO_WATCH)) {
                    if (flux_kvs_lookup_cancel (l->f) < 0)
                        flux_log_error (ctx->h, "%s: flux_kvs_lookup_cancel",
                                        __FUNCTION__);

                    if (flux_respond_error (ctx->h, l->msg, ENODATA, NULL) < 0)
                        flux_log_error (ctx->h, "%s: flux_respond_error",
                                        __FUNCTION__);
                }
                lookup_ctx_destroy (l);
            }
```

i.e. cancel the kvs-watcher watcher and send ENODATA to the current eventlog watcher on teardown.  

Should the latter return ENODATA?  It's true no more watch data is coming, but it's not returning ENODATA under a "good" or "normal" scenario, it's ending b/c of a "bad" situation.

It doesn't appear we send ENODATA on the equivalent situation in kvs-watch.

garlick commented:
Agreed it does seem like a different error is called for in that case

On Tue, Mar 12, 2019, 10:54 PM Codecov <notifications@github.com> wrote:

> Codecov
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071?src=pr&el=h1>
> Report
>
> Merging #2071
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071?src=pr&el=desc>
> into master
> <https://codecov.io/gh/flux-framework/flux-core/commit/c59f78a2569fcf331a244fd5f012367de1d8b735?src=pr&el=desc>
> will *decrease* coverage by 0.09%.
> The diff coverage is 73.84%.
>
> @@            Coverage Diff            @@
> ##           master    #2071     +/-   ##
>
> =========================================
> - Coverage   80.48%   80.38%   -0.1%
>
> =========================================
>
>   Files         192      193      +1
>
>   Lines       30246    30548    +302
>
> =========================================
> + Hits        24342    24557    +215
> - Misses       5904     5991     +87
>
> Impacted Files
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071?src=pr&el=tree> Coverage
> Δ
> src/common/libjob/job.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NvbW1vbi9saWJqb2Ivam9iLmM=> 74.78%
> <100%> (+8.88%) ⬆️
> src/modules/job-info/job-info.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL21vZHVsZXMvam9iLWluZm8vam9iLWluZm8uYw==> 68%
> <68%> (ø)
> src/cmd/flux-job.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NtZC9mbHV4LWpvYi5j> 84.49%
> <76.04%> (-3.49%) ⬇️
> src/common/libflux/ev_flux.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NvbW1vbi9saWJmbHV4L2V2X2ZsdXguYw==> 97.56%
> <0%> (-2.44%) ⬇️
> src/common/libflux/msg_handler.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NvbW1vbi9saWJmbHV4L21zZ19oYW5kbGVyLmM=> 89.06%
> <0%> (-1.18%) ⬇️
> src/connectors/local/local.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2Nvbm5lY3RvcnMvbG9jYWwvbG9jYWwuYw==> 88.57%
> <0%> (-0.72%) ⬇️
> src/modules/connector-local/local.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL21vZHVsZXMvY29ubmVjdG9yLWxvY2FsL2xvY2FsLmM=> 73.33%
> <0%> (-0.3%) ⬇️
> src/common/libflux/message.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NvbW1vbi9saWJmbHV4L21lc3NhZ2UuYw==> 81.39%
> <0%> (-0.25%) ⬇️
> src/cmd/flux-module.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NtZC9mbHV4LW1vZHVsZS5j> 83.72%
> <0%> (-0.24%) ⬇️
> src/common/libflux/reactor.c
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree#diff-c3JjL2NvbW1vbi9saWJmbHV4L3JlYWN0b3IuYw==> 91.85%
> <0%> (-0.23%) ⬇️
> ... and 5 more
> <https://codecov.io/gh/flux-framework/flux-core/pull/2071/diff?src=pr&el=tree-more>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2071#issuecomment-472289959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKX24ORjraVsI_FxTk638Nwnz3LA8X7ks5vWJKsgaJpZM4bpVRW>
> .
>

garlick commented:
Couple other nits (sorry missed earlier):
* In flux-job.c there is a function `parse_arg_unsigned()` which should be called to parse job id's rather than calling strtoull directly in the new subcommands.
* Similarly, there is an `iso_timestr()` function (msec precision) that could be used in place of `eventlog_timestr()` (usec precision).  I don't think usec precision is really needed in the eventlog output.
* It might be useful for `wait-event` to have an option to display the event (or just make that the default?)  In the tests anyway, we often wait for an event, then fetch the eventlog to extract the triggering event.

grondo commented:
> Similarly, there is an iso_timestr() function (msec precision) that could be used in place of eventlog_timestr() (usec precision). I don't think usec precision is really needed in the eventlog output.

I was actually going to suggest that the timestamps be left alone in seconds with usec precision. To me the ISO timestamp in UTC is not all that useful for viewing the eventlog for jobs -- e.g. you can't subtract two entries to get their difference, and events especially in debug mode may actually be ~1ms or less apart.

I think all the `eventlog` commands should:
 - display the unmodified timestamp by default (e.g. like Linux `dmesg(1)`)
 - have an option to display the ISO timestr *in current timezone* (e.g. like Linux `dmesg --time-format=iso`)
 - have an option to display the offset time (e.g. first entry is 0.0000, other entries are offsets)

garlick commented:
That sounds good to me @grondo.  What do you think @chu11?

grondo commented:
Of course, adding TZ support to `iso_timestr` and the support for relative timestamps do not need to land in this PR. I would prefer this PR pass the timestamps unmodified in `flux job eventlog` (i.e. make that the default, but just my opinion), then the other stuff could come in a future PR.

I'll open an issue.

> It might be useful for wait-event to have an option to display the event (or just make that the default?) In the tests anyway, we often wait for an event, then fetch the eventlog to extract the triggering event.

BTW, I agree this would be awesome! `flux job wait-event --quiet` could be used to wait without printing the entry from the matching event.. (and maybe `--verbose` could print even non-matching entries?)

chu11 commented:
Hey guys, I simply yoinked the timestr output from `flux kvs` without thinking about it.  I think all of the time output functions are a good idea, but agree it should be in a future PR.

chu11 commented:
diff coverage isn't what I'd like it to be, the main missed part is the cleanup done when a module is unloaded AND watchers still exist.  I'll write up something special to cover that to get atleast above 75%.

chu11 commented:
just re-pushed with fixes

- use `parse_arg_unsigned` appropriately
- `wait-event` outputs the event by default
- `--quiet` & `--verbose` options added for wait-event
- module unload corner case, returns ENOSYS instead of ENODATA.  I felt this was the right one, indicating the module is no longer loaded and the watch is no longer available.
- `eventlog` & `wait-event` output raw timestamp
  - (I didn't do this in flux-kvs, should I?)
- update `log_err` to `log_msg` in flux kvs given prior comments

garlick commented:
Nice!

Did notice one bug - `flux job wait-event --verbose` seems to dump events that occur after the triggering event when it's documented to (and probably should) dump only those leading up to and including it.  Example:
```
$ flux job wait-event -v 40617639936  submit 
1552515114.064075 submit userid=5588 priority=16 flags=0
1552515152.285945 alloc rank0/core[0-1]
```

grondo commented:
I gave the recent stuff a spin and it works great!

chu11 commented:
> occur after the triggering event when it's documented to (and probably should) dump only those leading up to and including it.

Hmmm.  I thought I handled that, but clearly I didn't do it correctly.

garlick commented:
I think this is probably ready to merge once we get that little bug fixed.

chu11 commented:
> Hmmm. I thought I handled that, but clearly I didn't do it correctly.

Doh! Had a corner case.  And I realize how a lot of the tests can be made simpler.  Re-pushed with fix and a bunch of test cleanups.

garlick commented:
Maybe a chainlint failure?  I've gotta run but I'll check in later and merge if this comes together.  Thanks!

grondo commented:
> Maybe a chainlint failure? 

In case you don't find it quickly, from travis logs:

```
PASS: t2204-job-info.t 7 - flux job eventlog fails on bad id
error: bug in the test script: broken &&-chain: 
        jobid=$(flux job submit test.json)
        flux job wait-event $jobid submit > wait_event1.out &&
        grep submit wait_event1.out
```

chu11 commented:
yeah, multiple chain lint issues.  Fixing up.

grondo commented:
Great! Looks like @garlick already approved and this is *super* useful, so merging!
